irep2 consistency, part 1/n

fbrausse opened the pull request:
This PR introduces a few invariants in the construction of expr2t in line with #204:
- ~`constant_struct2t` checks consistency of provided member expressions with the struct type~ (*not yet*)
- `with2t` checks consistency of the updated value with its source, as well as that of the field expression
- `arith_ops2` enforces some relations between the resulting type and also its 2 operands (this required to derive `shl2t` and `ashr2t` from `bit_ops2` instead of `arith_ops2`)
- `member2t` enforces that the source is a struct or union and that it actually contains the member
- ~`array_type2t` enforces a non-nil `size` expression to be of unsigned type~ (*not yet*)

~This is still based on #615, for now, and not ready for review at least until that PR is dealt with.~

fbrausse committed:
[irep2] add disabled consistency check for with2t

fbrausse committed:
[irep2] hide check behind NDEBUG

fbrausse committed:
[irep2] enable with2t check

fbrausse committed:
[irep2] make shl2t and ashr2t bit_ops2 instead of arith_op2

fbrausse committed:
[irep2] let arith_ops2 assert some consistency conditions

fbrausse committed:
[regression] fix for failure to link newly parsed elibc

fbrausse committed:
[irep2] non-cheri-fix: {address => pointer}_width

fbrausse committed:
clang-format

fbrausse committed:
[util] simplify expr2: use common type for new add2t

fbrausse committed:
[irep2] relaxed consistency cond for with2t: accomodate code, pointer and string types

fbrausse committed:
[util] simplify add2t: rework common-type handling by casting

fbrausse committed:
clang-format ... is this really necessary?

fbrausse commented:
- regression/esbmc/32_floppy fails a new consistency-assert because it multiplies a signed 32-bit array-size with an unsigned 64-bit size of the subtype in dereferencet::construct_struct_ref_from_dyn_offs_rec(). I don't see signed array sizes having any valid use case in expr2; is there any? Enforcing unsigned array sizes breaks in a lot of other regression tests though, would need to investigate. Should unsigned array sizes be part of this PR?
- regression/esbmc-unix/10_bicycle_04 fails on macOS because a value of type `struct timespec` is updated in symex-assign, but the type is broken/incomplete - it has no members. How can this happen?
  ```
  p lhs->dump()
  member
  * source_value : symbol
      * name : c:main.c@1964@F@cycling_thread@time_to_sleep
      * renamelev : Level 1
      * level1_num : 1
      * level2_num : 0
      * thread_num : 1
      * node_num : 0
      * type : struct
        * members : 
  
        * member_names : 
  
        * member_pretty_names : 
  
        * typename : struct timespec
        * packed : false
  * member_name : tv_sec
  * type : signedbv
    * width : 64
  ```
  however clang -E (with a [macOS-11.3 sysroot](https://github.com/phracker/MacOSX-SDKs.git)) produces the right definition which is not used for `time_to_sleep`:
  ```c
  # 33 "../../MacOSX-SDKs/MacOSX11.3.sdk/usr/include/sys/_types/_timespec.h" 3 4
  struct timespec
  {
   __darwin_time_t tv_sec;
   long tv_nsec;
  };
  ```

fbrausse committed:
[elibc] pthread.h needs to forward-decl "struct timespec", see #619

fbrausse commented:
> the type is broken/incomplete - it has no members. How can this happen? 

Never mind, got it: it happened because `struct timespec` was not declared globally before it was used in our <pthread.h>. Interesting, but consistent. Fixed.

fbrausse committed:
[irep2] member2t consistency: check that member exists in type

lucasccordeiro commented:
> Should unsigned array sizes be part of this PR?

@fbrausse: could you please create another PR to deal with this issue?

fbrausse committed:
[util] remove duplicate "if" and fun-decls

fbrausse committed:
[util] put c_link class into anon namespace

fbrausse committed:
[irep2] align gen_(zero|one)() error handling

fbrausse committed:
[regression] replace an incompatibly pre-proc'ed .i with .c from SV-COMP'22

This is c/pthread-atomic/time_var_mutex.c from SV-COMP ad265d07a04053bfcd80626ac1d30cec2d1c254e

fbrausse commented:
I have traced the incompatibly pre-processed regression/esbmc-unix/04_bzip2smp_01 back to a1d16931ae6a2351158cfcf14a9eb5bdbede07ce. Does anyone know what the original source of these "esmbc-pt" benchmarks was?

fbrausse committed:
[regression] de-preprocess esbmc-unix/04_bzip2smp_01

This restores compatibility with the pthread types defined in
the internal libc.

fbrausse committed:
[dereference] fix for signed array_size: type-cast before multiplying

fbrausse committed:
formatting

fbrausse committed:
[util] simplify conditions for arith_op2 type-casting

lucasccordeiro commented on src/irep2/irep2_expr.cpp:
```diff
@@ -350,6 +350,68 @@ expr2tc constant_string2t::to_array() const
   return final_val;
 }
 
+#ifndef NDEBUG
```
Do we really need this? If so, can you add some comments for this?

fbrausse replied:
This one we don't need, I've removed it.

lucasccordeiro commented on src/irep2/irep2_expr.h:
```diff
@@ -499,6 +499,29 @@ class arith_2ops : public arith_ops
     const expr2tc &v2)
     : arith_ops(t, id), side_1(v1), side_2(v2)
   {
+#ifndef NDEBUG
```
The same here. I would recommend adding some comments.

fbrausse replied:
I would keep this in the `.h`, otherwise it'll slow down Release-builds unnecessarily.

fbrausse replied:
I've added a comment.

lucasccordeiro commented on src/irep2/irep2_expr.h:
```diff
@@ -2807,6 +2832,9 @@ class with2t : public with_expr_methods
     const expr2tc &value)
     : with_expr_methods(type, with_id, source, field, value)
   {
+#if !defined(NDEBUG)
```
Comments?

lucasccordeiro commented on src/irep2/irep2_expr.h:
```diff
@@ -2827,6 +2855,15 @@ class member2t : public member_expr_methods
   member2t(const type2tc &type, const expr2tc &source, const irep_idt &memb)
     : member_expr_methods(type, member_id, source, memb)
   {
+#ifndef NDEBUG
```
Comments?

lucasccordeiro commented on src/pointer-analysis/dereference.cpp:
```diff
@@ -1262,7 +1262,7 @@ void dereferencet::construct_from_const_struct_offset(
     }
 
-    if(int_offset < (m_offs + m_size))
+    if(int_offset < m_offs + m_size)
```
I would prefer the previous version for the readability's sake :-)

fbrausse replied:
strange :) reverted.

lucasccordeiro reviewed (approved):
Minor comments only.

fbrausse committed:
[irep2] remove unnecessary #ifndef NDEBUG

fbrausse committed:
[dereference] restore parenthesis

https://github.com/esbmc/esbmc/pull/619#discussion_r811031019

fbrausse committed:
[irep2] explain #ifndef NDEBUG

fbrausse commented:
SV-COMP'22 with a "Release" build (i.e. no assertions) on master (5cce74ed0381c0795ae14eda2742dfb93ea63536):
```
Statistics:          15645 Files
  correct:            6591
    correct true:     3548
    correct false:    3043
  incorrect:            30
    incorrect true:      5
    incorrect false:    25
  unknown:            9024
  Score:              9579 (max: 25561)
```
This PR (33258c5eb6d3b8ddc6c4c05ac56b29f2f872977f):
```
Statistics:          15645 Files
  correct:            6601
    correct true:     3551
    correct false:    3050
  incorrect:            30
    incorrect true:      5
    incorrect false:    25
  unknown:            9014
  Score:              9592 (max: 25561)
```

lucasccordeiro commented:
Many thanks, @fbrausse.
